### PR TITLE
Detection Algorithm for Optimized Slasher

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

Now that we have all the required parts of slashing detection using the optimized approach from #8331, we need to glue it together into a cohesive algorithm to detect surrounding and surround votes. This PR begins the detection process, with an algorithm that looks like this when given a batch of attestations

1. Group the attestations by chunk index.
2. Update the min and max spans for those grouped attestations, check if any slashings are found in the process
3. Update the latest written epoch for all validators involved to the current epoch.

The algorithm for updating min and max spans is as follows:

1. Determine the chunks we need to use for updating for the validator indices in a validator chunk index, then retrieve those chunks from the database.
2. Using the chunks from step (1):

```
     for every attestation by chunk index:
        for each validator in the attestation's attesting indices:
          check if the attestation is slashable, if so return a slashing object.
```

3. Save the updated chunks to disk.

**Which issues(s) does this PR fix?**

Part of #8331

### CODE REVIEW

File: beacon-chain/db/iface/interface.go [hunk 1]
```diff
@@ -58,15 +58,15 @@ type ReadOnlyDatabase interface {
 	// Powchain operations.
```
Comment (rauljordan) [inline beacon-chain/db/iface/interface.go:58]: Indeed, batched reads and writes are the way to go here, so the interface methods are adapted to reflect that improvement

File: beacon-chain/db/kv/slasher.go [hunk 2]
```diff
@@ -6,35 +6,44 @@ import (
 	"fmt"
```
Comment (rauljordan) [inline beacon-chain/db/kv/slasher.go:6]: This file simply converts methods that were `single_input -> single_output` to `multiple_inputs -> multiple_outputs` for more efficient use of bolt for batched read/writes

File: beacon-chain/slasher/chunks.go [hunk 3]
```diff
@@ -87,11 +87,11 @@ type MaxSpanChunksSlice struct {
 // EmptyMinSpanChunksSlice initializes a min span chunk of length C*K for
 // C = chunkSize and K = validatorChunkSize filled with neutral elements.
 // For min spans, the neutral element is `undefined`, represented by MaxUint16.
-func EmptyMinSpanChunksSlice(config *Parameters) *MinSpanChunksSlice {
+func EmptyMinSpanChunksSlice(params *Parameters) *MinSpanChunksSlice {
```
Comment (rauljordan) [inline beacon-chain/slasher/chunks.go:90]: Input args renamed for consistency

File: beacon-chain/slasher/detect_attestations.go [hunk 4]
```diff
@@ -5,74 +5,175 @@ import (
...
+// as the current epoch in time, we perform slashing detection.
+// The process is as follows given a list of attestations:
+//
+// 1. Group the attestations by chunk index.
+// 2. Update the min and max spans for those grouped attestations, check if any slashings are
+//    found in the process
+// 4. Update the latest written epoch for all validators involved to the current epoch.
+//
+// This function performs a lot of critical actions and is split into smaller helpers for cleanliness.
+func (s *Service) detectSlashableAttestations(
```
Comment (rauljordan) [inline beacon-chain/slasher/detect_attestations.go:32]: Right now this function is kind of just gluing together certain methods, but it is not yet complete until next PR

File: beacon-chain/slasher/detect_attestations.go [hunk 5]
```diff
@@ -5,74 +5,175 @@ import (
...
+		return slashertypes.NotSlashable, err
+	}
+	// Load the required chunks from disk.
+	chunksByChunkIdx, err := s.loadChunks(ctx, opts, chunkIndices)
+	if err != nil {
+		return slashertypes.NotSlashable, err
+	}
+
+	// Apply the attestations to the related chunks.
+	// TODO(#8331): Apply...
```
Comment (rauljordan) [inline beacon-chain/slasher/detect_attestations.go:84]: The actual logic of applying an attestation to the min and max spans will be done in a separate PR

File: beacon-chain/slasher/helpers.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
+package slasher
```
Comment (rauljordan) [inline beacon-chain/slasher/helpers.go:1]: Helpers file created to add various pure functions used in slashing detection to a single, concise place

File: beacon-chain/slasher/helpers_test.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
+package slasher
```
Comment (rauljordan) [inline beacon-chain/slasher/helpers_test.go:1]: This is simply moved here from detect_attestations_test.go, no new logic nor tests

File: beacon-chain/slasher/receive_test.go [hunk 8]
```diff
@@ -112,96 +112,3 @@ func TestSlasher_receiveAttestations_OnlyValidAttestations(t *testing.T) {
 	}
 	require.DeepEqual(t, wanted, s.attestationQueue)
 }
-
```
Comment (rauljordan) [inline beacon-chain/slasher/receive_test.go:115]: Moved to helper_test.go

**Review by terencechain — APPROVED** [review]
